plugins/trace_api_plugin/include/eosio/trace_api_plugin/store_provider.hpp

```diff
@@ -0,0 +1,186 @@
+         :std::runtime_error(what_arg)
+      {}
+      explicit old_slice_version(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class incompatible_slice_files : public std::runtime_error {
+   public:
+      explicit incompatible_slice_files(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit incompatible_slice_files(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class malformed_slice_file : public std::runtime_error {
+   public:
+      explicit malformed_slice_file(const char* what_arg)
+         :std::runtime_error(what_arg)
+      {}
+      explicit malformed_slice_file(const std::string& what_arg)
+         :std::runtime_error(what_arg)
+      {}
+   };
+
+   class store_provider;
+
+
+   class slice_provider {
+   public:
+      struct index_header {
+         uint32_t version;
+      };
+      slice_provider(const boost::filesystem::path& slice_dir, uint32_t width);
+
+      uint32_t slice_number(uint32_t block_height) const {
+         return block_height / _width;
+      }
+
+      bool find_index_slice(uint32_t slice_number, bool append, fc::cfile& slice_file) const;
+      bool find_trace_slice(uint32_t slice_number, bool append, fc::cfile& slice_file) const;
+
+   private:
+      // returns true if slice is created
+      bool find_slice(const char* slice_prefix, uint32_t slice_number, fc::cfile& slice_file) const;
+
+      const boost::filesystem::path _slice_dir;
+      const uint32_t _width;
+
+      static constexpr uint32_t _current_version = 1;
+
+      static constexpr uint _slice_size = 10;
+      static constexpr uint _null_term_size = 1;
+      static constexpr const char* _trace_prefix = "trace_";
+      static constexpr const char* _trace_index_prefix = "trace_index_";
+      static constexpr uint _max_file_size = 12 + 10 + 1 + 10 + 4 + 1; // "trace_index_" + 10-digits + '-' + 10-digits + ".log" + null-char
+   };
+
+   class store_provider {
+   public:
+      store_provider(const boost::filesystem::path& slice_dir, uint32_t width);
+
+      void append(const block_trace_v0& bt);
+      void append_lib(uint32_t lib);
+
+      /**
+       * Read the metadata log font-to-back starting at an offset passing each entry to a provided functor/lambda
+       *
+       * @tparam Fn : type of the functor/lambda
+       * @param block_height : height of the requested data
+       * @param offset : initial offset to read from
+       * @param fn : the functor/lambda
+       * @return the highest offset read during this scan
+       */
+      template<typename Fn>
+      uint64_t scan_metadata_log_from( uint32_t block_height, uint64_t offset, Fn&& fn ) {
+         // ignoring offset
+         offset = 0;
+         fc::cfile index;
+         const uint32_t slice_number = _slice_provider.slice_number(block_height);
+         _slice_provider.find_index_slice(slice_number, false, index);
```

Comment:
A find method should never create. Maybe rename to `find_or_create_index_slice`

Reply:
Moreover, the `scan_metadata_log_from` is a reader function and the intent is that the reading process _never_ mutates the filesystem.  I would prefer a hard line in the code where we can reasonably assure ourselves that this will never create a file.